File: tests/basic/struct_set_explicit.ml
open Optitrust
open Target


let _ = Run.doc_script_cpp (fun _ ->

    !! Struct_basic.set_explicit [sInstr "a = b"];

)

"
typedef struct {
  int x;
  int y;
} vect;

int main() {
  vect a;
  vect b;
  a = b;
}
"
let _ = Flags.dump_ast_details := true


let _ = Run.script_cpp ( fun _ ->

  !! Struct_basic.set_explicit [sInstr "b = p"];
  !! Struct_basic.set_explicit [sInstr "u = a.pos"];
  !! Struct_basic.set_explicit [sInstr "t[0] = p2"];
  !! Struct_basic.set_explicit [sInstr "c = a"];
  !! Struct_basic.set_explicit [sInstr "c.pos ="];
  !! Struct_basic.set_explicit [sInstr "c.speed ="];


  !! Trace.alternative (fun () ->
    !! Struct_basic.set_explicit [sInstr "c = a"];
    !! Struct_basic.set_explicit [nbMulti;cOr [[sInstr "c.pos ="]; [sInstr "c.speed ="]]];
    !! ());
)

File: tests/basic/struct_set_explicit.cpp

typedef struct {
  int x;
  int y; }
vect;

typedef struct {
  int weight;
  vect pos;
  vect speed;
} obj;

vect f() {
  return {1,1};
}

int main() {
  int x = 1;
  int y;
  x = y;
  vect p = {0,0};
  vect b;
  b = p;
  vect u;
  obj a = {0,{0,0},{0,0}};
  u = a.pos;
  vect t[2];
  vect p2 = p;
  t[0] = p2;
  obj c;
  c = a;
  return 0;
}
